### ISSUE TITLE
reconsider taking HandeRef as input parameters when the parameter must be a handle of some sort.

### ISSUE THREAD

**Issue by AraHaan** [microsoft/CsWin32#202]
Currently in CsWin32 no overloads taking in ``HandleRef``s are generated.

The use cases:
- ``HandleRef`` is not only used by user applications (for referencing ``HWND`` handles to (most commonly) gui controls to pass them into p/invoke methods to the operating system apis).
- They are also used in the runtime itself (they are used a lot in the dotnet organization on github).

This is where my code uses it: https://github.com/Elskom/Els_kom_new/search?q=HandleRef&type=code (it seems to not show every usage of it for some reason in my NativeMethods.cs file.)

Got 616 results for it as well too https://github.com/search?q=org%3Adotnet+HandleRef&type=code, a lot of commits, and issues regarding it as well.

More results could be in the code bases within all of the repositories in the dotnet organization on github, one way to find out for sure how many results come up is to clone them all, and then within Visual Studio search for ``HandleRef`` in every last repository cloned from there to get the absolute number of usages in the code within that organization.

I also encourage an in general look to see how many other people might also use ``HandleRef`` as well too.

As seen above, ``HandleRef`` while documented as deprecated in .NET 2.0, is still widely used everywhere even within the runtime itself with valid use cases that make it impossible to use say SafeHandle, expecially when you own the object that made the handle but not the handle itself and as such using ``HandleRef``-or ``GC.KeepAlive(objectmadeinthiscode);`` are the only options to force the GC to not finalize the object (with said handle), before the native operating system calls finish with it and possibly cause undefined behavior on user's applications (and possibly the runtime itself as well).

This is why one or the other should be done, either tell the user to ``tell the GC to keep that object alive.`` or use a ``HandeRef`` that would need to be generated by CsWin32.

Another option for this, is to have it not generate ``HandleRef`` overloads by default but have the user opt-in for them if they specifically ask for said overload to be generated within the configuration for specific members only (opt-in per p/invoke function name) within the text file that controls the source generator.

There was more discussion about this in #125.

**Comment by AArnott** [comment]
Use within the runtime itself is not an argument that applies to this repo as the runtime repo is not our target audience.
Use by the public API that user applications must interact with is a potential reason. Do those exist?

We need a good reason to generate what was deprecated so long ago. Can you explain a scenario where `HandleRef` does the job better than `SafeHandle`?

**Comment by jnm2** [comment]
@AArnott Again, this is for scenarios where you do not have control of the code that created the handle, such as `System.Windows.Forms.Form`. HandleRef keeps you from forgetting to do `GC.KeepAlive(form);` after the native call that uses the form's handle.

**Comment by AArnott** [comment]
> I do not know if the Windows Forms or WPF teams would welcome the use of undocumented windows apis being used inside of those libraries despite it now being easy to be able to do.

I'm sure they wouldn't, on legal grounds. Microsoft products outside Windows cannot call into undocumented Windows APIs.

> SafeHandle is never going to help with safety here because it doesn't prevent the object from being concurrently collected

Why do you say that? Who is going to finalize a `SafeHandle` while it's in a p/invoke call? That's what its addref/releaseref methods are for. SafeHandle should never close the native handle while it's in use. Even if you explicitly Dispose of it, I _think_ it waits till the last ReleaseRef call before closing the native handle.

> HandleRef keeps you from forgetting to do GC.KeepAlive(form);

Well, `form` is not a native handle, and we would not generate a HandleRef or a SafeHandle for a `Form` class so I'm not sure how that's relevant.

> for scenarios where you do not have control of the code that created the handle, such as System.Windows.Forms.Form.

`Form` doesn't derive from `HandleRef`, so I'm again not sure what you're talking about.

If WinForms has public APIs that produce HandleRef for you to use and you want to use that in CsWin32-generated APIs, then that's a more solid case. If those APIs exist, can you produce examples?

**Comment by jnm2** [comment]
@AArnott I must have not been communicating clearly. I'm not sure how to reply to most of your responses.

Here's the situation that HandleRef is useful for, and there's simply no way that I can ever use SafeHandle to help prevent `Form` from being finalized and freeing its handle:

```cs
void DoSomething(Form form)
{
    // Danger! The GC might finalize 'form' during this call, freeing the handle.
    User32.EnumChildWindows(form.Handle, ..., ...);
}
```

HandleRef pressures you towards success by not forgetting to keep 'form' alive.
```cs
void DoSomething(Form form)
{
    User32.EnumChildWindows(new HandleRef(form, form.Handle), ..., ...);
}
```

SafeHandle makes no sense here because the `Form` object that you may or may not own has its own finalizer which frees its own IntPtr handle, which you decidedly don't own.

```cs
void DoSomething(Form form)
{
    // Nothing is stopping the GC from concurrently finalizing 'form' and freeing the handle during this call
    User32.EnumChildWindows(new SomeKindOfSafeHandle(form.Handle), ..., ...);
}
```

**Comment by jnm2** [comment]
You could say "the problem is clearly that Windows Forms and similar libraries should be using SafeHandle," but that doesn't solve the question of what to do with *today's* .NET 3.0–6, and .NET Framework which will probably never deprecate its IntPtr handle property even if .NET 10 does.

**Comment by jnm2** [comment]
Let me know if the examples aren't explaining enough for you to see why it's critical to have GC.KeepAlive after a call that uses a handle that can be freed by the finalizer of a managed non-SafeHandle object that you don't own the source code for. It's bitten me many times.

**Comment by jnm2** [comment]
Yes. Now that I'm thinking about it, you'd probably need to expose an analyzer because the source generator probably can't itself be the one to register for IArgumentOperations etc to check and provide warnings for all call sites.

**Comment by jnm2** [comment]
Oh, also, separate analyzer would be analyzing a semantic model of the compilation that already has the generated source, which probably makes everything easier than dealing with missing types and members.
